Enable using Kokkos with CUDA backend for MIRCO and in general (#2012)
4C-multiphysics/4C: doc/documentation/src/installation/installation.rst

MIRCO can be used as optional dependency inside |FOURC| to be used for linear elastic frictionless normal contact between a rigid rough indentor and an elastic half-space.
See the `MIRCO repository <https://github.com/imcs-compsim/MIRCO>`_ for details and downloads.

> Building |FOURC| with MIRCO enabled automatically fetches the repository during the configure stage and later builds the library as dependency. Alternatively, one can specify an external MIRCO installation. In either case, MIRCO can make use of shared memory parallelism through Kokkos :ref:`when enabled <build4Cwithkokkoscuda>` in |FOURC|. Note that 4C and MIRCO must depend on the same Kokkos installation. In case using Kokkos with CUDA enabled, MIRCO must be built with `CMAKE_POSITION_INDEPENDENT_CODE=ON`.

Reviewer: Writing down details about TPLs is bound to be outdated quickly.

Reviewer 2: Question for understanding: is the position independent code flag needed on the MICRO or 4C side or both?

> Building |FOURC| with MIRCO enabled automatically fetches the repository during the configure stage and later builds the library as dependency. Alternatively, an external MIRCO installation can be specified. In either case, MIRCO can make use of shared memory parallelism through Kokkos :ref:`when enabled <build4Cwithkokkosparallel>` in |FOURC|. Note that 4C and MIRCO must depend on the same Kokkos installation. In case using Kokkos with CUDA enabled, MIRCO must be built with `CMAKE_POSITION_INDEPENDENT_CODE=ON`.

Reviewer 2: ```suggestion
Building |FOURC| with MIRCO enabled automatically fetches the repository during the configure stage and later builds the library as dependency. Alternatively, an external MIRCO installation can be specified. In either case, MIRCO can make use of shared memory parallelism through Kokkos :ref:`when enabled <build4Cwithkokkosparallel>` in |FOURC|. Note that 4C and MIRCO must depend on the same Kokkos installation. When using Kokkos with CUDA enabled, MIRCO must be built with `CMAKE_POSITION_INDEPENDENT_CODE=ON`.
```

    # This pulls in all the necessary dependencies and headers.
    target_link_libraries(<your-target> PRIVATE 4C::lib4C)

.. _build4Cwithkokkoscuda:

Building |FOURC| with OpenMP and CUDA support through Kokkos
------------------------------------------------------------

|FOURC| is primarily developed around MPI parallelism, but also offers the ability to use shared memory parallelism through `Kokkos <https://kokkos.org/> _`, enabling hybrid parallelism on the CPU through OpenMP and GPU acceleration through CUDA.

> Kokkos (and Kokkos-Kernels) can be built within Trilinos or specified as an external TPL in Trilinos, and its configuration follows the usual procedure for the desired backend (see the `Kokkos configuration guide <https://kokkos.org/kokkos-core-wiki/get-started/configuration-guide.html> _`). Trilinos then requires `Trilinos_ENABLE_<Backend>=ON` and, specifically for CUDA, `Trilinos_ENABLE_TPL_CUDA=ON`. To prevent oversubscription and unwanted shared memory parallelism in 4C, one should disable these backends for Tpetra with `TPETRA_INST_<BACKEND>=OFF` and explicitly set `TPETRA_INST_SERIAL=ON`.

Reviewer 2: Question: Does this potentially induce a bottleneck that could noticeably slow end to end simulation time? I don't know where/how Tpetra is used or how much of the end to end time it typically consumes, so this is for my curiosity and trying to think ahead.

Reviewer 3: For now, this is not a bottleneck, since we cannot use Tpetra anyways. The need set `TPETRA_INST_<BACKEND>=OFF` and explicitly set `TPETRA_INST_SERIAL=ON` is just, that we are still using Epetra. This will hopefully change, when we move to Tpetra at some point in the future.

> To build 4C with this configuration, a compiler wrapper, `utilities/clangcuda++` must be used as the `CMAKE_CXX_COMPILER`, while clang should be used as the `CMAKE_C_COMPILER`. When using MPI, these should instead be set as the `OMPI_CXX` and `OMPI_CC` environment variables respectively. To change the GPU architecture or default clang++ and CUDA paths, one should set the corresponding environment variables listed at the start of the `utilities/clangcuda++` compiler wrapper. Additionally, the `FOUR_C_CLANGCUDA` compile option must be enabled in 4C. Due to incompatibility with the serial version of ArborX, it is recommended to disable `FOUR_C_WITH_ARBORX`.

Reviewer 2: Question: 'should' or 'must' for using Clang as the C compiler? Can one use GCC for C while using Clang for CXX? Has this combination been tested with 4C?

Also, 'clang' or 'Clang' here?

Reviewer 2: Additional note for this section - it might be worth mentioning to users that given the large number of dependencies, failure late in the building process is normal so setup for 4C with CUDA enabled Kokkos may be very time consuming

    # This pulls in all the necessary dependencies and headers.
    target_link_libraries(<your-target> PRIVATE 4C::lib4C)

.. _build4Cwithkokkoscuda:

Building |FOURC| with OpenMP and CUDA support through Kokkos
------------------------------------------------------------

|FOURC| is primarily developed around MPI parallelism, but also offers the ability to use shared memory parallelism through `Kokkos <https://kokkos.org/> _`, enabling hybrid parallelism on the CPU through OpenMP and GPU acceleration through CUDA.

> Kokkos (and Kokkos-Kernels) can be built within Trilinos or specified as an external TPL in Trilinos, and its configuration follows the usual procedure for the desired backend (see the `Kokkos configuration guide <https://kokkos.org/kokkos-core-wiki/get-started/configuration-guide.html> _`). Trilinos then requires `Trilinos_ENABLE_<Backend>=ON` and, specifically for CUDA, `Trilinos_ENABLE_TPL_CUDA=ON`. To prevent oversubscription and unwanted shared memory parallelism in 4C, one should disable these backends for Tpetra with `TPETRA_INST_<BACKEND>=OFF` and explicitly set `TPETRA_INST_SERIAL=ON`.

Reviewer 2: Note - you say here that you need `Trilinos_ENABLE_TPL_CUDA=ON`, but you do not have that in the script for Docker above. You instead have `TPL_ENABLE_CUDA=ON`:
```
  -D Tpetra_INST_SERIAL:BOOL=ON \
  \
  -D Trilinos_ENABLE_CUDA:BOOL=ON \
  -D TPL_ENABLE_CUDA:BOOL=ON \
```

Reviewer 2: Trilinos itself says that we want what is in the script for Docker above. https://mperego.github.io/trilinos.github.io/building_with_CUDA_support.html

This does highlight the risk in copying pieces of TPL documentation into here. We assume the responsibility for the correctness, which I don't think we want.

> To build 4C with this configuration, a compiler wrapper, `utilities/clangcuda++` must be used as the `CMAKE_CXX_COMPILER`, while clang must be used as the `CMAKE_C_COMPILER`. When using MPI, these must instead be set as the `OMPI_CXX` and `OMPI_CC` environment variables respectively. To change the GPU architecture or default clang++ and CUDA paths, one should set the corresponding environment variables listed at the start of the `utilities/clangcuda++` compiler wrapper. Additionally, the `FOUR_C_CLANGCUDA` compile option must be enabled in 4C. Due to incompatibility with the serial version of ArborX, it is recommended to disable `FOUR_C_WITH_ARBORX`.

Reviewer 2: This says

> When using MPI, these must instead be set as the `OMPI_CXX` and `OMPI_CC` environment variables respectively. 

which implies that `CMAKE_CXX_COMPILER` and `CMAKE_C_COMPILER` should not be set as in the previous sentence - but that leads to a compilation error for me.

I still cannot get this to build after a week of trying, so either I am missing something in these directions or these directions are missing something, or both. I think at the very least *I* am misunderstanding something here.

Reviewer 3: @PhilipOesterlePekrun, I do not know where the problem comes from. Though, it is detrimental that others can build this, before we merge.

Reviewer 2: hmm, I think sometimes `CLANGCUDA_MODE_DEVICE` is being set when I would not expect that to happen:
```
FAILED: src/core/fem/benchmark_tests/CMakeFiles/benchmarktests_core.dir/__/__/__/__/tests/benchmark_tests/4C_benchmark_tests_main.cpp.o 
/home/thompson/Dev/4C/utilities/clangcuda++ -DBOOST_GRAPH_DYN_LINK -DBOOST_GRAPH_NO_LIB -DBOOST_MAJOR_VERSION=1 -DBOOST_MINOR_VERSION=83 -DBOOST_REGEX_DYN_LINK -DBOOST_REGEX_NO_LIB -DCLANGCUDA_MODE_DEVICE -DKOKKOS_DEPENDENCE -DMIRCO_ENABLE_VISUALIZATIONEXPORT=0 -DMPICH_SKIP_MPICXX -DOMPI_SKIP_MPICXX -D_MPICC_H -I/home/thompson/Dev/4C/src -I/home/thompson/Dev/4C/src/adapter    .....    -O3 -DNDEBUG -std=c++20 -fPIE -Wall -Wextra -Wvla -Wno-unused-parameter -Wno-overloaded-virtual -Werror -ftrapping-math -O3 -funroll-loops -Wno-unknown-pragmas -extended-lambda -Wext-lambda-captures-this -expt-relaxed-constexpr -arch=sm_86 -MD -MT src/core/fem/benchmark_tests/CMakeFiles/benchmarktests_core.dir/__/__/__/__/tests/benchmark_tests/4C_benchmark_tests_main.cpp.o -MF src/core/fem/benchmark_tests/CMakeFiles/benchmarktests_core.dir/__/__/__/__/tests/benchmark_tests/4C_benchmark_tests_main.cpp.o.d -o src/core/fem/benchmark_tests/CMakeFiles/benchmarktests_core.dir/__/__/__/__/tests/benchmark_tests/4C_benchmark_tests_main.cpp.o -c /home/thompson/Dev/4C/tests/benchmark_tests/4C_benchmark_tests_main.cpp
```

I'm not sure what I'm doing that is making the build system think that `/home/thompson/Dev/4C/tests/benchmark_tests/4C_benchmark_tests_main.cpp` should have any reason to be compiled for the device :thinking:

Reviewer 2: Ok, I ended up using CUDA 13.3 + Clang 23. This seems to work now with the following changes:

- Set the benchmark tests to CLANGCUDA_MODE_HOST
- Fix https://github.com/4C-multiphysics/4C/pull/2121
- add `[[maybe_unused]]` for two parameters in `src/fluid_turbulence/4C_fluid_turbulence_statistic_manager.hpp` and `src/porofluid_pressure_based/4C_porofluid_pressure_based_meshtying_strategy_artery.hpp`

Was this tested with `ninja full`? I couldn't get the documentation to build, so I suspect that something needs to be changed in CMake files to pass `-DCLANGCUDA_MODE_HOST` to those flies, as they definitely don't run on the GPU.

I say 'seems to work' because I do get one test failure:
```console
$ ctest -L minimal
Test project /home/thompson/Dev/4C/build/cuda

 ...

      Start 2971: unittests_contact_constitutivelaw.threads2
25/44 Test #2971: unittests_contact_constitutivelaw.threads2 ...........................................***Failed    3.84 sec

...

98% tests passed, 1 tests failed out of 44

Label Time Summary:
minimal    = 151.01 sec*proc (43 tests)

Total Test time (real) =  87.02 sec

The following tests did not run:
	2748 - beam3r_line2_genalpha_liegroup_contact_penalty_linpen_twocrossedbeams.4C.yaml-p2 (Skipped)

The following tests FAILED:
	2971 - unittests_contact_constitutivelaw.threads2 (Failed) minimal
```
and
```console
$ ctest -R unittests
Test project /home/thompson/Dev/4C/build/cuda

...

      Start 2971: unittests_contact_constitutivelaw.threads2
 4/22 Test #2971: unittests_contact_constitutivelaw.threads2 ........***Failed    3.84 sec

...

95% tests passed, 1 tests failed out of 22

Label Time Summary:
minimal    =  62.28 sec*proc (22 tests)

Total Test time (real) =  45.99 sec

The following tests FAILED:
	2971 - unittests_contact_constitutivelaw.threads2 (Failed) minimal
```

Reviewer 3: Both reported tests are actually the same test. 

As far as I remember, this tests executes MIRCO under the hood, so this is really at the core of this PR. I just checked the pipeline and there it seems, that the test passes (https://github.com/4C-multiphysics/4C/actions/runs/29146341477/job/86528807510?pr=2012). 

@PhilipOesterlePekrun Any thoughts on this?

Reviewer 2: Digging into the verbose output, it looks like Kokkos is failing to launch a GPU kernel, so yeah, this isn't a successful build

Reviewer 2: ```console
$ nvidia-smi
Tue Jul 14 15:01:02 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 610.43.02              KMD Version: 610.43.02     CUDA UMD Version: 13.3     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA RTX A1000 6GB Lap...    On  |   00000000:01:00.0 Off |                  N/A |
| N/A   46C    P0            312W /   35W |      14MiB /   6144MiB |      6%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A            3297      G   /usr/lib/xorg/Xorg                        4MiB |
+-----------------------------------------------------------------------------------------+
```
```console
$ ctest -R unittests --rerun-failed --output-on-failure
Test project /home/thompson/Dev/4C/build/cuda
    Start 2971: unittests_contact_constitutivelaw.threads2
1/1 Test #2971: unittests_contact_constitutivelaw.threads2 ...***Failed    5.66 sec
[==========] Running 13 tests from 7 test suites.
[----------] Global test environment set-up.
[----------] 1 test from FourCGoogleTest
[ RUN      ] FourCGoogleTest.RunningWithIntendedNumberOfProcesses
[       OK ] FourCGoogleTest.RunningWithIntendedNumberOfProcesses (0 ms)
[----------] 1 test from FourCGoogleTest (0 ms total)

[----------] 2 tests from BrokenrationalConstitutiveLawTest
[ RUN      ] BrokenrationalConstitutiveLawTest.TestEvaluate
[       OK ] BrokenrationalConstitutiveLawTest.TestEvaluate (0 ms)
[ RUN      ] BrokenrationalConstitutiveLawTest.TestEvaluateDeriv
[       OK ] BrokenrationalConstitutiveLawTest.TestEvaluateDeriv (0 ms)
[----------] 2 tests from BrokenrationalConstitutiveLawTest (0 ms total)

[----------] 2 tests from CubicConstitutiveLawTest
[ RUN      ] CubicConstitutiveLawTest.TestEvaluate
[       OK ] CubicConstitutiveLawTest.TestEvaluate (0 ms)
[ RUN      ] CubicConstitutiveLawTest.TestEvaluateDeriv
[       OK ] CubicConstitutiveLawTest.TestEvaluateDeriv (0 ms)
[----------] 2 tests from CubicConstitutiveLawTest (0 ms total)

[----------] 2 tests from LinearConstitutiveLawTest
[ RUN      ] LinearConstitutiveLawTest.TestEvaluate
[       OK ] LinearConstitutiveLawTest.TestEvaluate (0 ms)
[ RUN      ] LinearConstitutiveLawTest.TestEvaluateDeriv
[       OK ] LinearConstitutiveLawTest.TestEvaluateDeriv (0 ms)
[----------] 2 tests from LinearConstitutiveLawTest (0 ms total)

[----------] 2 tests from MircoConstitutiveLawForceTest
[ RUN      ] MircoConstitutiveLawForceTest.TestEvaluate
unknown file: Failure
C++ exception with description "(cuda_instance->cuda_func_get_attributes_wrapper(&attr, func)) error( cudaErrorNoKernelImageForDevice): no kernel image is available for execution on the device /home/thompson/Dev/4C-dependencies/tk_cuda/include/kokkos/Cuda/Kokkos_Cuda_KernelLaunch.hpp:141" thrown in the test fixture's constructor.

[  FAILED  ] MircoConstitutiveLawForceTest.TestEvaluate (0 ms)
[ RUN      ] MircoConstitutiveLawForceTest.TestEvaluateDeriv
unknown file: Failure
C++ exception with description "(cuda_instance->cuda_func_get_attributes_wrapper(&attr, func)) error( cudaErrorNoKernelImageForDevice): no kernel image is available for execution on the device /home/thompson/Dev/4C-dependencies/tk_cuda/include/kokkos/Cuda/Kokkos_Cuda_KernelLaunch.hpp:141" thrown in the test fixture's constructor.

[  FAILED  ] MircoConstitutiveLawForceTest.TestEvaluateDeriv (0 ms)
[----------] 2 tests from MircoConstitutiveLawForceTest (0 ms total)

[----------] 2 tests from MircoConstitutiveLawPressureTest
[ RUN      ] MircoConstitutiveLawPressureTest.TestEvaluate
unknown file: Failure
C++ exception with description "(cuda_instance->cuda_func_get_attributes_wrapper(&attr, func)) error( cudaErrorNoKernelImageForDevice): no kernel image is available for execution on the device /home/thompson/Dev/4C-dependencies/tk_cuda/include/kokkos/Cuda/Kokkos_Cuda_KernelLaunch.hpp:141" thrown in the test fixture's constructor.

[  FAILED  ] MircoConstitutiveLawPressureTest.TestEvaluate (0 ms)
[ RUN      ] MircoConstitutiveLawPressureTest.TestEvaluateDeriv
unknown file: Failure
C++ exception with description "(cuda_instance->cuda_func_get_attributes_wrapper(&attr, func)) error( cudaErrorNoKernelImageForDevice): no kernel image is available for execution on the device /home/thompson/Dev/4C-dependencies/tk_cuda/include/kokkos/Cuda/Kokkos_Cuda_KernelLaunch.hpp:141" thrown in the test fixture's constructor.

[  FAILED  ] MircoConstitutiveLawPressureTest.TestEvaluateDeriv (0 ms)
[----------] 2 tests from MircoConstitutiveLawPressureTest (0 ms total)

[----------] 2 tests from PowerConstitutiveLawTest
[ RUN      ] PowerConstitutiveLawTest.TestEvaluate
[       OK ] PowerConstitutiveLawTest.TestEvaluate (0 ms)
[ RUN      ] PowerConstitutiveLawTest.TestEvaluateDeriv
[       OK ] PowerConstitutiveLawTest.TestEvaluateDeriv (0 ms)
[----------] 2 tests from PowerConstitutiveLawTest (0 ms total)

[----------] Global test environment tear-down
[==========] 13 tests from 7 test suites ran. (0 ms total)
[  PASSED  ] 9 tests.
[  FAILED  ] 4 tests, listed below:
[  FAILED  ] MircoConstitutiveLawForceTest.TestEvaluate
[  FAILED  ] MircoConstitutiveLawForceTest.TestEvaluateDeriv
[  FAILED  ] MircoConstitutiveLawPressureTest.TestEvaluate
[  FAILED  ] MircoConstitutiveLawPressureTest.TestEvaluateDeriv

 4 FAILED TESTS
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[55765,1],0]
  Exit code:    1
```
everything should be supported, but something is wrong for sure. I wonder if `sm_86` is being set correctly here? I really am not sure, but at least I've got new errors now that everything compiles. Well, everything but the documentation, but that's not the issue at hand here.

Reviewer 2: I could just do my development in the Docker container? It would unblock me but is not a typically something you want to rely upon

Reviewer 2: Update - as a insanity check, I verified that my previous work using raw CUDA works fine on this machine with my current Clang and CUDA and drivers, so the problem is somewhere in the heavy 4C+Trilinos+Kokkos setup

Reviewer 2: I'm investigating a possible theory that my env and config variable of sm_86 is being ignored in favor of sm_90, leading to the issue of incompatible kernels

Reviewer 2: Yup, I changed `clangcuda++` in both Micro and 4C to default to sm_86 and now that test passes.... so perhaps those values are not passing along correctly when compiling Kokkos code. As a test, I'd recommend that someone else tries setting the default in `clangcuda++` to an arch their hardware cannot use and see if their environment variables are similarly not picked up. If so, then you should see an error like the one I had. If that happens for someone else as well, then we've got issues properly passing around values through the dependency stack.

Reviewer 2: Ok, so the problem is that my `.bashrc` had a bug, so `CLANGCUDA_ARCH` and `CUDA_ARCH` were not correctly being exported.

So this begs the question, *should* we be assuming an SM to target? When we do that, we end up silently emitting code that may not work on the user's hardware. We *do* want to support cross compilation, but that should be intentional rather than accidental, imho. Two possible fixes here

1) Error if `CLANGCUDA_ARCH` is not found

2) Attempt to determine an appropriate setting if `CLANGCUDA_ARCH` is not found

I have a possible approach to 2) given below

(note, a couple of other fixes in there as well, but those are covered in #2121 and #2127)

```diff
$ git diff
diff --git a/cmake/functions/four_c_auto_define_benchmark_tests.cmake b/cmake/functions/four_c_auto_define_benchmark_tests.cmake
index 8af9b98914..60eb30bff9 100644
--- a/cmake/functions/four_c_auto_define_benchmark_tests.cmake
+++ b/cmake/functions/four_c_auto_define_benchmark_tests.cmake
@@ -31,7 +31,7 @@ function(_set_up_benchmark_test_target _module_under_test _target)
   target_link_libraries(${_target} PRIVATE unittests_common)
 
   if(FOUR_C_CLANGCUDA)
-    set_clangcuda_mode(${_target} CLANGCUDA_MODE_DEVICE)
+    set_clangcuda_mode(${_target} CLANGCUDA_MODE_HOST)
   endif()
 
   # configure the respective input test with the respective mesh
diff --git a/src/cut/4C_cut_kernel.hpp b/src/cut/4C_cut_kernel.hpp
index 87aecd3e00..e1d356f903 100644
--- a/src/cut/4C_cut_kernel.hpp
+++ b/src/cut/4C_cut_kernel.hpp
@@ -2200,7 +2200,7 @@ namespace Cut::Kernel
         if (det < 1.0e-16) std::cout << "!!! determinant of jacobian is smaller than 1.0e-16 !!!\n";
       }
 
-      if (det < 1.0e-16 && det > 1.0e-16)
+      if (det < 1.0e-16 && det > -1.0e-16)
 #if EXTENDED_CUT_DEBUG_OUTPUT
         std::cout << "Determinant in  compute position is very close to zero" << std::endl;
 #endif
diff --git a/src/fluid_turbulence/4C_fluid_turbulence_statistic_manager.hpp b/src/fluid_turbulence/4C_fluid_turbulence_statistic_manager.hpp
index b020df2959..ead82c3c24 100644
--- a/src/fluid_turbulence/4C_fluid_turbulence_statistic_manager.hpp
+++ b/src/fluid_turbulence/4C_fluid_turbulence_statistic_manager.hpp
@@ -166,7 +166,7 @@ namespace FLD
     void setup();
 
     //! time step size
-    double dt_;
+    [[maybe_unused]] double dt_;
 
     //! parameters for sampling/dumping period
     //! start of sampling at step samstart
diff --git a/src/porofluid_pressure_based/4C_porofluid_pressure_based_meshtying_strategy_artery.hpp b/src/porofluid_pressure_based/4C_porofluid_pressure_based_meshtying_strategy_artery.hpp
index 247e86068c..cb04bea402 100644
--- a/src/porofluid_pressure_based/4C_porofluid_pressure_based_meshtying_strategy_artery.hpp
+++ b/src/porofluid_pressure_based/4C_porofluid_pressure_based_meshtying_strategy_artery.hpp
@@ -115,7 +115,7 @@ namespace PoroPressureBased
     PorofluidAlgorithm* porofluid_algorithm_;
 
     //! global parameter list of the algorithm
-    const Teuchos::ParameterList& params_;
+    [[maybe_unused]] const Teuchos::ParameterList& params_;
 
     //! parameter list of porofluid algorithm
     const Teuchos::ParameterList& porofluid_params_;
diff --git a/utilities/clangcuda++ b/utilities/clangcuda++
index 39e3681b1d..a24ca52d67 100755
--- a/utilities/clangcuda++
+++ b/utilities/clangcuda++
@@ -23,7 +23,12 @@ cuda_path="${CLANGCUDA_CUDA_PATH:-${CUDA_HOME:-/usr/local/cuda}}"
 compile=0
 mode_cuda_host=0
 mode_cuda_device=0
-arch="${CLANGCUDA_ARCH:-sm_90}"
+arch_flag=()
+if [[ -n "${CLANGCUDA_ARCH}" ]]; then
+  arch_flag=(--cuda-gpu-arch="${CLANGCUDA_ARCH}")
+else
+  arch_flag=(--cuda-gpu-arch="sm_$(nvidia-smi --query-gpu=compute_cap --format=csv,noheader|head -n 1 | sed "s/[.]//g")")
+fi
 
 args=()
 skip_next_x=0
@@ -99,7 +104,7 @@ if [[ "$compile" == "1" && "$mode_cuda_host" == "1" ]]; then
     -x cuda
     --cuda-host-only
     --cuda-path="$cuda_path"
-    --cuda-gpu-arch="$arch"
+    "${arch_flag[@]}"
     -Wno-unknown-cuda-version
     "${args[@]}"
   )
@@ -108,7 +113,7 @@ elif [[ "$compile" == "1" && "$mode_cuda_device" == "1" ]]; then
     "$clang_path"
     -x cuda
     --cuda-path="$cuda_path"
-    --cuda-gpu-arch="$arch"
+    "${arch_flag[@]}"
     -Wno-unknown-cuda-version
     "${args[@]}"
   )
```

Reviewer 2: Really, we should also dump out a nice error message if `CLANGCUDA_ARCH` was not set (and maybe also `CUDA_ARCH`), and the query to `nvidia-smi` failed to return a valid #.#

This was just a first hack at a way to auto-detect the user hardware if they didn't set the `CLANGCUDA_ARCH`... but I'm still not quite sure if we want to do that for the user or not.

> To build 4C for the CUDA backend, enable the `FOUR_C_CLANGCUDA` compile option. Additionally, a compiler wrapper, `utilities/clangcuda++`, must be used as the C++ compiler and clang must be used as the C compiler. For an MPI build (`-DCMAKE_CXX_COMPILER=</path/to/mpicxx>`; `-DCMAKE_C_COMPILER=</path/to/mpicc>`), you should set the following environment variables: `export OMPI_CXX=</path/to/4C>/utilities/clangcuda++` and `export OMPI_CC=</path/to/clang>`. The compiler wrapper should detect and use the same architecture as the Kokkos installation, but, in case of errors or cross-compilation, you can explicitly override this by setting the `FOUR_C_CLANGCUDA` environment variable. Similarly, the default clang++ and CUDA paths can be overridden with the environment variables `CLANGCUDA_CLANG_PATH` and `CLANGCUDA_CUDA_PATH`, respectively.

Reviewer 4: Can you add a brief explanation of why clang must be used? You already gave a reason in the PR description. This would make it clearer why we use clang. In the future, the reason why clang must be used may disappear, and we can remove this requirement instead of holding on to it for an unknown reason.

> Building 4C for the CUDA backend currently necessitates the use of a compiler wrapper, `utilities/clangcuda++`, as the C++ compiler and clang as the C compiler. The reason for this is that Nvidia's NVCC (or the `kokkos_launch_compiler` which is based on NVCC) has issues compiling some modern C++ code in 4C, and using clang resolves these issues. To build 4C for the CUDA backend, you must first enable the `FOUR_C_CLANGCUDA` compile option. For an MPI build of 4C (`-DCMAKE_CXX_COMPILER=</path/to/mpicxx>`; `-DCMAKE_C_COMPILER=</path/to/mpicc>`), you should then set the following environment variables: `export OMPI_CXX=</path/to/4C>/utilities/clangcuda++` and `export OMPI_CC=</path/to/clang>`. The compiler wrapper will detect and use the same architecture as the Kokkos installation, but, in case of errors or cross-compilation, you can explicitly override this by setting the `CLANGCUDA_ARCH` environment variable. Similarly, the default clang++ and CUDA paths can be overridden with the environment variables `CLANGCUDA_CLANG_PATH` and `CLANGCUDA_CUDA_PATH`, respectively.

Reviewer 2: ```suggestion
Building 4C with Kokkos's CUDA backend enabled currently requires using a compiler wrapper, `utilities/clangcuda++`, as the C++ compiler and clang as the C compiler. Nvidia's NVCC, which Kokkos uses by default, has issues compiling some modern C++ code in 4C. To build 4C with Kokkos's CUDA backend enabled, you must first set the `FOUR_C_CLANGCUDA` compile option. For an MPI build of 4C, you should then set the following environment variables: `export OMPI_CXX=</path/to/4C>/utilities/clangcuda++` and `export OMPI_CC=</path/to/clang>`. The compiler wrapper will detect and use the same architecture as the Kokkos installation, but you can explicitly override this by setting the `CLANGCUDA_ARCH` environment variable. Similarly, the default clang++ and CUDA paths can be overridden with the environment variables `CLANGCUDA_CLANG_PATH` and `CLANGCUDA_CUDA_PATH`, respectively.
```
I would try to use simpler sentence structure overall. As academics we are trained to use cumbersome English sentence structures, but it really hurts readability and user understanding, imho (Also a problem in making papers that other researchers can understand and use if you ask me!)

|FOURC| is primarily developed around MPI parallelism, but also offers the ability to use shared memory parallelism through `Kokkos <https://kokkos.org/> _`, enabling hybrid parallelism on the CPU through OpenMP and GPU acceleration through CUDA.

> Kokkos (and Kokkos-Kernels) can be built within Trilinos or specified as an external TPL in Trilinos, and its configuration follows the usual procedure for the desired backend (see the `Kokkos configuration guide <https://kokkos.org/kokkos-core-wiki/get-started/configuration-guide.html> _`). Trilinos then requires `Trilinos_ENABLE_<Backend>=ON` and, specifically for CUDA, `TPL_ENABLE_CUDA=ON`. To prevent oversubscription and unwanted shared memory parallelism in 4C, one should disable these backends for Tpetra with `TPETRA_INST_<BACKEND>=OFF` and explicitly set `TPETRA_INST_SERIAL=ON`.

Reviewer 2: note - this change is now not consistent with the commit message for the commit that contains it. Not the end of the world, but makes navigating the commit history, by human eyes or by LLM, more difficult. 🤷
